File: 0295-find-median-from-data-stream/0295-find-median-from-data-stream.py
import heapq
class MedianFinder:

    def __init__(self):
        self.total=0
        self.minHeapForLargeNums=[]
        self.maxHeapForSmallNums=[]
    def addNum(self, num: int) -> None:
        if len(self.maxHeapForSmallNums)==0 or len(self.minHeapForLargeNums)+1 > len(self.maxHeapForSmallNums):
            heapq.heappush(self.maxHeapForSmallNums, -num)
        else:
            heapq.heappush(self.minHeapForLargeNums, num)
        
        if (len(self.maxHeapForSmallNums)!=0 and len(self.minHeapForLargeNums)!=0):
            while -self.maxHeapForSmallNums[0]>self.minHeapForLargeNums[0]:
                ele=heapq.heappop(self.maxHeapForSmallNums)
                ele1=heapq.heappop(self.minHeapForLargeNums)
                heapq.heappush(self.maxHeapForSmallNums,-ele1)
                heapq.heappush(self.minHeapForLargeNums,-ele)

        self.total+=num

    def findMedian(self) -> float:
        if (len(self.maxHeapForSmallNums)+ len(self.minHeapForLargeNums))%2==0:
            return (-self.maxHeapForSmallNums[0]+self.minHeapForLargeNums[0])/2
            # return self.total/(len(self.maxHeapForSmallNums)+len(self.minHeapForLargeNums))
        else:
            return -self.maxHeapForSmallNums[0]


# Your MedianFinder object will be instantiated and called as such:
# obj = MedianFinder()
# obj.addNum(num)
# param_2 = obj.findMedian()

File: 0295-find-median-from-data-stream/0295-find-median-from-data-stream.java
class MedianFinder {

    PriorityQueue<Integer> minHeapForLargeNums;
    PriorityQueue<Integer> maxHeapForSmallNums;

    public MedianFinder() {
        maxHeapForSmallNums = new PriorityQueue<>(Collections.reverseOrder());
        minHeapForLargeNums = new PriorityQueue<>();
    }

    public void addNum(int num) {
            PriorityQueue<Integer> minHeapForLargeNums1=minHeapForLargeNums;
            PriorityQueue<Integer> maxHeapForSmallNums1=maxHeapForSmallNums;
        if (maxHeapForSmallNums.size() == 0 || maxHeapForSmallNums.peek() >= num) {
            maxHeapForSmallNums.offer(num);
            if (maxHeapForSmallNums.size() > minHeapForLargeNums.size()+ 1 ) {
                minHeapForLargeNums.offer(maxHeapForSmallNums.poll());
            }
        } else {
            minHeapForLargeNums.offer(num);
            if (minHeapForLargeNums.size() > maxHeapForSmallNums.size()) {
                maxHeapForSmallNums.offer(minHeapForLargeNums.poll());
            }
        }

    }

    public double findMedian() {
        int total = minHeapForLargeNums.size() + maxHeapForSmallNums.size();
        if (total == 0) {
            return 0;
        }
        int b = minHeapForLargeNums.size() != 0 ? minHeapForLargeNums.peek() : 0;
        int a = maxHeapForSmallNums.size() != 0 ? maxHeapForSmallNums.peek() : 0;
        if (total % 2 == 0) {
            return (a + b) / (double) 2;
        } else {
            return a;
        }
    }
}

/**
 * Your MedianFinder object will be instantiated and called as such:
 * MedianFinder obj = new MedianFinder();
 * obj.addNum(num);
 * double param_2 = obj.findMedian();
 */